src/stores/WalletStore.ts

```diff
@@ -38,24 +37,95 @@ class WalletStore extends PersistableStore {
     return web3Modal.cachedProvider
   }
 
-  async connect(clearCachedProvider = false) {
+  private async getUserNetworkName() {
+    return (await provider.getNetwork()).name
+  }
+
+  private async isNetworkRight() {
+    const network = env.VITE_ETH_NETWORK
+    const userNetwork = await this.getUserNetworkName()
+    return network === userNetwork
+  }
+
+  private async handleAccountChanged() {
+    if (!provider) return
+
+    this.walletLoading = true
+    this.account = await this.getAccount()
+    NotificationsStore.showTwitterShare = false
+    this.walletLoading = false
+  }
+
+  private async getAccount() {
+    return (await provider.listAccounts())[0]
+  }
+
+  private clearData() {
+    NotificationsStore.showTwitterShare = false
+    web3Modal.clearCachedProvider()
+    this.account = undefined
+  }
+
+  private async switchNetwork() {
+    const chainId = hexValue(env.VITE_CHAIN_ID)
+    await provider.send('wallet_switchEthereumChain', [{ chainId }])
+  }
+
+  private subscribeProvider(provider: Web3Provider) {
+    if (!provider.on) return
+
+    provider.on('error', (error: Error) => {
+      handleError(error)
+    })
+
+    provider.on('accountsChanged', (accounts: string[]) => {
+      if (!accounts.length) this.clearData()
+
+      this.account = undefined
+      void this.handleAccountChanged()
+    })
+    provider.on('disconnect', (error: unknown) => {
+      if (isSwitchingNetworkError(error)) return handleError(error)
+
+      if (provider) provider.removeAllListeners()
+      handleError(error)
+      this.clearData()
+    })
+    provider.on('chainChanged', async () => {
+      if (await this.isNetworkRight()) {
+        this.needNetworkChange = false
+        this.account = await this.getAccount()
+        return
+      }
+      this.account = undefined
+      this.needNetworkChange = true
+    })
+  }
+
+  private async connect(clearCachedProvider = false) {
     this.walletLoading = true
     try {
+      if (provider) provider.removeAllListeners()
       if (clearCachedProvider) web3Modal.clearCachedProvider()
 
       const instance = await web3Modal.connect()
-      provider = new Web3Provider(instance)
-
-      await this.checkNetwork(provider)
-      if (this.needNetworkChange)
-        throw new Error(
-          ErrorList.wrongNetwork(
-            (await provider.getNetwork()).name,
-            env.VITE_ETH_NETWORK
+      // We need this "any" networks so we could handle when users switch between networks with listeners without errors
+      provider = new Web3Provider(instance, 'any')
+
+      this.account = await this.getAccount()
+      const isNetworkNotRight = !(await this.isNetworkRight())
+      if (isNetworkNotRight) {
+        this.needNetworkChange = true
```

Comment:
You have:
```TS
      this.account = undefined
      this.needNetworkChange = true
```
in `subscribeProvide ... provider.on('chainChanged')`, let's extract these two lines into `private` function